components/feature/accounts/src/main/java/mozilla/components/feature/accounts/WebChannelFeature.kt

```diff
@@ -0,0 +1,331 @@
+                WebChannelCommand.CAN_LINK_ACCOUNT -> processCanLinkAccountCommand(
+                    messageId, engineSession
+                )
+                WebChannelCommand.FXA_STATUS -> processFxaStatusCommand(
+                    accountManager, messageId, engineSession
+                )
+                WebChannelCommand.OAUTH_LOGIN -> processOauthLoginCommand(
+                    accountManager, payload
+                )
+            }
+        }
+    }
+
+    private fun registerContentMessageHandler(session: Session) {
+        val engineSession = sessionManager.getOrCreateEngineSession(session)
+        val messageHandler = WebChannelViewContentMessageHandler(engineSession, accountManager)
+        registerMessageHandler(engineSession, messageHandler)
+    }
+
+    @VisibleForTesting
+    companion object {
+        private val logger = Logger("mozac-fxawebchannel")
+
+        internal const val WEB_CHANNEL_EXTENSION_ID = "mozacWebchannel"
+        internal const val WEB_CHANNEL_EXTENSION_URL =
+            "resource://android/assets/extensions/fxawebchannel/"
+
+        // Constants for incoming messages from the WebExtension.
+        private const val CHANNEL_ID = "account_updates"
+
+        enum class WebChannelCommand {
+            CAN_LINK_ACCOUNT,
+            OAUTH_LOGIN,
+            FXA_STATUS
+        }
+
+        // For all possible messages and their meaning/payloads, see:
+        // https://github.com/mozilla/fxa/blob/master/packages/fxa-content-server/docs/relier-communication-protocols/fx-webchannel.md
+
+        /**
+         * Gets triggered when user initiates a login within FxA web content.
+         * Expects a response.
+         * On Fx Desktop, this event triggers "a different user was previously signed in on this machine" warning.
+         */
+        private const val COMMAND_CAN_LINK_ACCOUNT = "fxaccounts:can_link_account"
+
+        /**
+         * Gets triggered when a user successfully authenticates via OAuth.
+         */
+        private const val COMMAND_OAUTH_LOGIN = "fxaccounts:oauth_login"
+
+        /**
+         * Gets triggered on startup to fetch the FxA state from the host application.
+         * Expects a response, which includes application's capabilities and a description of the
+         * current Firefox Account (if present).
+         */
+        private const val COMMAND_STATUS = "fxaccounts:fxa_status"
+
+        @Volatile
+        internal var installedWebExt: WebExtension? = null
+
+        @Volatile
+        private var registerContentMessageHandler: (WebExtension) -> Unit? = { }
+
+        internal var ports = WeakHashMap<EngineSession, Port>()
+
+        /**
+         * Installs the WebChannel web extension in the provided engine.
+         *
+         * @param engine a reference to the application's browser engine.
+         */
+        fun install(engine: Engine) {
+            engine.installWebExtension(WEB_CHANNEL_EXTENSION_ID, WEB_CHANNEL_EXTENSION_URL,
```

Comment:
This had to bring in a lot of code from the reader view feature. I've filed https://github.com/mozilla-mobile/android-components/issues/4297 to address the code duplication and come up with an abstraction.